Fix RemoveDeploymentTriggerDecorator order

Sgitario opened the pull request:
The decorator RemoveDeploymentTriggerDecorator was not configured to be triggered before ChangeDeploymentTriggerDecorator, so sometimes the image stream tag was wrong.

This change fixes the failure spotted by the test OpenshiftWithDockerAndImageTest where the image stream tag was sometimes 0.1-SNAPSHOT and other times 1.0 when it should be always 1.0.

Relates to the random failures in https://github.com/dekorateio/dekorate/actions/runs/4023231521/jobs/6913879180 and also in the jakarta-rewrite branch

iocanel commented on extensions/kubernetes/vanilla/deployment/src/main/java/io/quarkus/kubernetes/deployment/RemoveDeploymentTriggerDecorator.java:
```diff
@@ -16,7 +15,7 @@ public void andThenVisit(DeploymentConfigSpecFluent<?> deploymentConfigSpec, Obj
     }
 
     @Override
-    public Class<? extends Decorator>[] after() {
```
This should not be removed.

The purpose of `RemoveDeploymentTrigger` is to remove triggers either provided by the user or by the `ApplyDeploymentTrigger` decorator. By removing this condition it's possible that `RemoveDeploymentTrigger` runs before there is anything to remove.

iocanel reviewed (changes requested):
There is an issue with the implementation as commented directly on the code.
Also this is fixed as part of #30566 so we could possibly close this one.

Sgitario commented:
> There is an issue with the implementation as commented directly on the code. Also this is fixed as part of #30566 so we could possibly close this one.

Unfortunately, this is not fixing the issue even using your changes because the ApplyDeploymentTriggerDecorator from decorator is being applied the last:

```
[INFO] Applying decorator 'io.quarkus.kubernetes.deployment.RemoveDeploymentTriggerDecorator' -- removes the image triggers if any
[INFO] Applying decorator 'io.quarkus.kubernetes.deployment.ChangeDeploymentTriggerDecorator' -- add the correct image
[INFO] Applying decorator 'io.dekorate.openshift.decorator.ApplyDeploymentTriggerDecorator' -- overwrite the wrong image
```

Sgitario commented:
@iocanel I've just updated the changes that seem to consistently apply the decorators in the correct order.
Also, I've followed your suggestion to keep using ApplyDeploymentTriggerDecorator in RemoveDeploymentTriggerDecorator. Additionally, I've updated the RemoveDeploymentTriggerDecorator decorator to only do that to the DeploymentConfig with the correct name.

Sgitario commented:
I'm running out of ideas here since locally my changes worked consistently ok, but not in CI.
Now, I'm doing exactly what was done by @iocanel in https://github.com/quarkusio/quarkus/pull/30566 in spite of that locally it was not working for me, but just in case CI accepts it.

Sgitario commented:
Something really odd is happening with the dekorate logic that orders the decorators to be run.
In spite of the CI being green, these changes are consistently not working on my local machine. On the other hand, the changes that indeed fix the actual issue locally do not work for the CI. so, we need further investigation here. Moving to Draft.

iocanel commented:
> Something really odd is happening with the dekorate logic that orders the decorators to be run. In spite of the CI being green, these changes are consistently not working on my local machine. On the other hand, the changes that indeed fix the actual issue locally do not work for the CI. so, we need further investigation here. Moving to Draft.

Are you sure that tests on CI are passing? I am seeing them as skipped.
Eitherway here are some thoughts on how we could deal with it:

- Simplify the original decorator in dekorate. The decorator seems to be quite complex due to a long fixed issue in sundrio, so we could possibly try to simplify it.
- Fix the ordering algorithm in dekorate. The issue here is that our comparator is not transitive which is most likely what is causing our ordering issues.
- Register a configurator that sets the right image so that no matter what decorator is run last it will use ther correct value.

Sgitario commented:
> > Something really odd is happening with the dekorate logic that orders the decorators to be run. In spite of the CI being green, these changes are consistently not working on my local machine. On the other hand, the changes that indeed fix the actual issue locally do not work for the CI. so, we need further investigation here. Moving to Draft.
> 
> Are you sure that tests on CI are passing? I am seeing them as skipped. Eitherway here are some thoughts on how we could deal with it:
> 
> * Simplify the original decorator in dekorate. The decorator seems to be quite complex due to a long fixed issue in sundrio, so we could possibly try to simplify it.
> * Fix the ordering algorithm in dekorate. The issue here is that our comparator is not transitive which is most likely what is causing our ordering issues.
> * Register a configurator that sets the right image so that no matter what decorator is run last it will use ther correct value.

+100 It makes totally sense to fix this from the root. 
I will try to use a custom configurator to do this.

Sgitario commented:
It seems that https://github.com/dekorateio/dekorate/issues/1147 is indeed a blocker because for some reason, it expects a build service factory (??) in Kind projects. 
So, I don't know how we cannot overwrite the default container image at the moment. @iocanel can you help here please?

Sgitario commented:
I'm going to confirm that this pull request https://github.com/quarkusio/quarkus/pull/30591 introduced this issue and if so, I will revert those changes.

Sgitario commented:
I've just confirmed that this issue has nothing to do with https://github.com/quarkusio/quarkus/pull/30591.
I guess this issue has been always there, and for some reason, now it's more frequent. The only fix I foresee is to fix the logic of the decorator execution at Dekorate side.

Sgitario committed:
Fix RemoveDeploymentTriggerDecorator order

iocanel commented:
> I've just confirmed that this issue has nothing to do with #30591. I guess this issue has been always there, and for some reason, now it's more frequent. The only fix I foresee is to fix the logic of the decorator execution at Dekorate side.

I think that issue was always there but only manifests itself under certain decorator combinations.
I agree that the best course of action, is to deal with the issue in the dekorate side. I'll be able to work on that later today.

Sgitario commented:
Closing as this will be addressed in Dekorate.
